### PR TITLE
Get token prices from our API instead of CoinGecko

### LINKED ISSUE #3574: [assets-controllers] Switch from CoinGecko to MetaMask price API
We are switching from CoinGecko to our own price API. This API can gather information from CoinGecko as well as other sources.

For further details, see this internal discussion thread: https://consensys.slack.com/archives/C065W3877E3/p1700678572843909?thread_ts=1700677118.264359&cid=C065W3877E3

### PR DESCRIPTION
**Pull request by mcmire**
## Explanation

CoinGecko is planning on retiring their `/simple/token_price` endpoint soon. This endpoint is critical for us, as we use to fetch prices for tokens. This commit removes all code related to CoinGecko in TokenRatesController so that it hits an internal API instead.

This is set up in such a way that we can upgrade to newer versions of the API (or even other services) in the future without breaking backward compatibility. However, at the moment it does introduce breaking changes to TokenRatesController.

## References

Fixes #3574.

## Changelog

(Updated in PR)

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+  // Metis Andromeda Mainnet
+  '0x440',
+  // Moonbeam
+  '0x504',
+  // Moonriver
+  '0x505',
+  // Base
+  '0x2105',
+  // Shiden
+  // NOTE: This is the wrong chain ID, this should be 0x150
```
Comment (mcmire) [inline packages/assets-controllers/src/token-prices-service/codefi-v2.ts:209]: I'm asking Tomas about this: https://github.com/MetaMask/metamask-mobile/pull/7957#discussion_r1412567673

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+type SpotPricesEndpointData<
+  TokenContractAddress extends Hex,
+  Currency extends string,
+> = Record<TokenContractAddress, Record<Currency, number>>;
+
+/**
+ * The list of currencies that can be supplied as the `vsCurrency` parameter to
+ * the `/spot-prices` endpoint, in lowercase form.
+ */
+export const SUPPORTED_CURRENCIES = [
```
Comment (mcmire) [inline packages/assets-controllers/src/token-prices-service/codefi-v2.ts:23]: This list comes from https://github.com/MetaMask/metamask-mobile/pull/7957.

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+type SupportedCurrency =
+  | (typeof SUPPORTED_CURRENCIES)[number]
+  | Uppercase<(typeof SUPPORTED_CURRENCIES)[number]>;
+
+/**
+ * The list of chain IDs that can be supplied as the `vsCurrency` parameter to
+ * the `/spot-prices` endpoint, but in hexadecimal form (for consistency with
+ * how we represent chain IDs in other places).
+ */
+export const SUPPORTED_CHAIN_IDS = [
```
Comment (mcmire) [inline packages/assets-controllers/src/token-prices-service/codefi-v2.ts:161]: This list comes from https://github.com/MetaMask/metamask-mobile/pull/7957.

**Review by mikesposito — DISMISSED** [review]
Looks good! Probably just some changes to do on inline docs

**Comment by mikesposito** [comment]
I am thinking about how/if this relates to [this issue](https://github.com/MetaMask/core/issues/3573).

Do we handle the case where a list of tokens overflows the service rate limit? And in case not, what would be the best way to do that after these changes?

**Comment by mcmire** [comment]
> Do we handle the case where a list of tokens overflows the service rate limit? And in case not, what would be the best way to do that after these changes?

I think we can wrap the call to `this.#tokenPricesService.fetchTokenPrices` so that we divide the list of tokens into batches, and then call that method for each batch to get the total set of token prices. Does that answer your question?

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+  // Aurora Mainnet
+  '0x4e454152',
+  // Harmony Mainnet Shard 0
+  '0x63564c40',
+] as const;
+
+/**
+ * A currency that can be supplied as the `vsCurrency` parameter to
+ * the `/spot-prices` endpoint. Covers both uppercase and lowercase versions.
+ */
```
[the remark this answers is not in the document]
Comment (mcmire) [reply packages/assets-controllers/src/token-prices-service/codefi-v2.ts]: Oops! This was clearly a bad copy/paste job. Fixed here: [cc6051f](https://github.com/MetaMask/core/pull/3600/commits/cc6051f4a86000716070740528f0072478eec926)

**Review by mikesposito — APPROVED** [review]

**Pull request merged**